IO/src/main/java/io/deephaven/io/log/LogEntry.java

```diff
@@ -93,54 +93,90 @@ default LogEntry appendDouble(Double f) {
     /**
      * Append a double to the given number of decimal places, rounding up.
      *
-     * @param f a double value to append to the logEntry
+     * @param doubleValue a double value to append to the logEntry
      * @param decimalPlaces a positive integer between 0 and 9
      * @return the resulting {@code LogEntry}
      */
-    default LogEntry appendPositiveDouble(double f, int decimalPlaces) {
+    default LogEntry appendDoubleToDecimalPlaces(final double doubleValue, final int decimalPlaces) {
+        return appendDoubleDecimalPlacesImpl(this, doubleValue, decimalPlaces, true);
+    }
+
+    /**
+     * Append a double with decimal places up to the given number.
+     *
+     * @param doubleValue a double value to append to the logEntry
+     * @param decimalPlaces a positive integer between 0 and 9
+     * @return the resulting {@code LogEntry}
+     */
+    default LogEntry appendDoubleToMaxDecimalPlaces(final double doubleValue, final int decimalPlaces) {
+        return appendDoubleDecimalPlacesImpl(this, doubleValue, decimalPlaces, false);
+    }
+
+    private static LogEntry appendDoubleDecimalPlacesImpl(
+            LogEntry entry, final double doubleValue, final int decimalPlaces, final boolean exact) {
         final int decimalPlacesAsPowerOf10 = MathUtil.pow10(decimalPlaces);
-        final long lf = (long) (0.5 + f * decimalPlacesAsPowerOf10);
-        final long integral = lf / decimalPlacesAsPowerOf10;
-        LogEntry r = append(integral);
-        if (decimalPlaces == 0) {
-            return r;
+        final boolean negative = doubleValue < 0.0;
+        final long longWeightedValue;
+        if (negative) {
+            longWeightedValue = -(long) (-0.5 + doubleValue * decimalPlacesAsPowerOf10);
+        } else {
+            longWeightedValue = (long) (0.5 + doubleValue * decimalPlacesAsPowerOf10);
+        }
+        final long integral = longWeightedValue / decimalPlacesAsPowerOf10;
+        if (negative) {
+            entry = entry.append(-integral);
+        } else {
+            entry = entry.append(integral);
         }
-        final int fractional = (int) (lf % decimalPlacesAsPowerOf10);
-        r = r.append(".");
+        if (decimalPlaces == 0) {
+            return entry;
+        }
+        int fractional = (int) (longWeightedValue % decimalPlacesAsPowerOf10);
+        int actualDecimalPlaces = decimalPlaces;
+        if (!exact) {
+            while (fractional > 0 && (fractional % 10 == 0)) {
+                fractional /= 10;
+                --actualDecimalPlaces;
+            }
+            if (fractional == 0) {
+                return entry;
+            }
+        }
+        entry = entry.append(".");
         final int base10FractionalDigits = MathUtil.base10digits(fractional);
-        final int leadingZeroes = decimalPlaces - base10FractionalDigits;
+        final int leadingZeroes = actualDecimalPlaces - base10FractionalDigits;
```

Comment:
What prevents someone from being a jerk and telling you they want 20 decimal places and then getting zero of them?